blockchain/message_pool/src/msgpool/mod.rs

```diff
@@ -1204,6 +1204,60 @@ where
     Ok(())
 }
 
+/// Like head_change, except it doesnt change the state of the MessagePool.
+/// It simulates a head change call.
+pub async fn run_head_change<T>(
+    api: &RwLock<T>,
+    pending: &RwLock<HashMap<Address, MsgSet>>,
+    from: Tipset,
+    to: Tipset,
+    rmsgs: &mut HashMap<Address, HashMap<u64, SignedMessage>>,
+) -> Result<(), Error>
+where
+    T: Provider + 'static,
+{
+    // TODO: This logic should probably be implemented in the ChainStore. It handles reorgs.
+    let mut left = Arc::new(from);
+    let mut right = Arc::new(to);
+    let mut left_chain = Vec::new();
+    let mut right_chain = Vec::new();
+    while left != right {
+        if left.epoch() > right.epoch() {
+            left_chain.push(left.as_ref().clone());
+            let par = api.read().await.load_tipset(left.parents()).await?;
+            left = par;
+        } else {
+            right_chain.push(right.as_ref().clone());
+            let par = api.read().await.load_tipset(right.parents()).await?;
+            right = par;
+        }
+    }
+    for ts in left_chain {
+        let mut msgs: Vec<SignedMessage> = Vec::new();
```

Comment:
You could reserve a capacity here

Reply:
Could I really? I would need to know how many messages there are in each block... Dont think i can know that without going through each block first.

Reply:
Oh my bad, you indeed don't have that info